packages/grid/_modules_/grid/hooks/features/export/useGridPrintExport.tsx

```diff
@@ -0,0 +1,277 @@
+        printDoc.body.classList.add(...normalizeOptions.bodyClass.split(' '));
+      }
+
+      if (normalizeOptions.copyStyles) {
+        const headStyleElements = doc.current!.querySelectorAll("style, link[rel='stylesheet']");
+
+        for (let i = 0; i < headStyleElements.length; i += 1) {
+          const node = headStyleElements[i];
+          if (node.tagName === 'STYLE') {
+            const newHeadStyleElements = printDoc.createElement(node.tagName);
+            const sheet = (node as HTMLStyleElement).sheet;
+
+            if (sheet) {
+              let styleCSS = '';
+              // NOTE: for-of is not supported by IE
+              for (let j = 0; j < sheet.cssRules.length; j += 1) {
+                if (typeof sheet.cssRules[j].cssText === 'string') {
+                  styleCSS += `${sheet.cssRules[j].cssText}\r\n`;
+                }
+              }
+              newHeadStyleElements.appendChild(printDoc.createTextNode(styleCSS));
+              printDoc.head.appendChild(newHeadStyleElements);
+            }
+          } else if (node.getAttribute('href')) {
+            // If `href` tag is empty, avoid loading these links
+
+            const newHeadStyleElements = printDoc.createElement(node.tagName);
+
+            for (let j = 0; j < node.attributes.length; j += 1) {
+              const attr = node.attributes[j];
+              if (attr) {
+                newHeadStyleElements.setAttribute(attr.nodeName, attr.nodeValue || '');
+              }
+            }
+
+            printDoc.head.appendChild(newHeadStyleElements);
+          }
+        }
+      }
+
+      // Trigger print
+      printWindow.contentWindow!.print();
+    },
+    [apiRef, doc, visibleRowCount, rowHeight, headerHeight],
+  );
+
+  const handlePrintWindowAfterPrint = React.useCallback(
+    (printWindow: HTMLIFrameElement): void => {
+      // Remove the print iframe
+      doc.current!.body.removeChild(printWindow);
+
+      // Revert grid to previous state
+      setGridState((state) => ({
+        ...state,
+        ...previousGridState.current,
+      }));
+
+      // Revert columns to their original state
+      if (previousHiddenColumns.current.length) {
+        apiRef.current.updateColumns(
+          columns.map((column) => {
+            column.hide = previousHiddenColumns.current.includes(column.field);
+            return column;
+          }),
+        );
+      }
+
+      // Clear local state
+      previousGridState.current = null;
+      previousHiddenColumns.current = [];
+    },
+    [columns, apiRef, setGridState],
+  );
+
+  const exportDataAsPrint = React.useCallback(
+    async (options?: GridPrintExportOptions) => {
+      logger.debug(`Export data as Print`);
+
+      if (!apiRef.current.rootElementRef!.current) {
+        throw new Error('No Grid Root element available');
+      }
+
+      previousGridState.current = gridState;
+
+      if (props.pagination) {
+        apiRef.current.setPageSize(visibleRowCount);
```

Comment:
This won't work if the user is controlling the page size. Forcing the state also doesn't work, because it will require a `forceUpdate()` to propagate the changes, and when this runs, all effects will also run again and reset the value back to the prop value. I see some options:

1. Add a escape hatch to temporarily disable the state update constraint. https://github.com/mui-org/material-ui-x/blob/95fed89847c5f1489ddb63dafedcf0d907d46354/packages/grid/_modules_/grid/hooks/features/core/useGridControlState.ts#L40
2. Add a note to the docs mentioning that the print export doesn't work if `pageSize` or `page` is present.
3. Render the grid inside the iframe and disable the virtualization: 

```jsx
ReactDOM.render(<DataGrid {...props} disableVirtualization />, iframeEl)
```

There're going to be a dependency cycle with 3., which I believe we could solve by using `require()` instead of `import`.

Reply:
I would go with option 2. -> the reason is that 3. for me will be too expensive in terms of the time it will take to render.
Also, this feature (along with a couple of other ones) doesn't work with server-side pagination which is controlled. If a developer is controlling the pagination the print export has an API that they can call manually.

This looks to me like a corner case - we can see what the feedback will be and if it is problem then we can spend time fixing it.